ncm-spma/src/main/perl/spma/apt.pm

```diff
@@ -0,0 +1,433 @@
+        my $prots = $repo->{protocols}->[0];
+        my $fh = CAF::FileWriter->new("$repos_dir/$repo->{name}.list", log => $self);
+        print $fh "# File generated by ", __PACKAGE__, ". Do not edit\n";
+        if (!exists($repo->{proxy}) && @px) {
+            if ($type eq 'reverse') {
+                $repo->{protocols} = $self->generate_reverse_proxy_urls($repo->{protocols}, @px);
+            } else {
+                $repo->{proxy} = $self->generate_forward_proxy_urls($repo->{protocols}, @px);
+            }
+        }
+        my $rs = $self->template()->process($template, $repo, $fh);
+        if (!$rs) {
+            $self->error ("Unable to generate repository $repo->{name}: ", $self->template()->error());
+            $fh->cancel();
+            return 0;
+        }
+        $changes += $fh->close() || 0; # handle undef
+        $fh->close();
+    }
+
+    return $changes;
+}
+
+
+# Returns a set of all installed packages
+sub get_installed_pkgs
+{
+    my $self = shift;
+    $self->debug(5, 'Entered installed_pkgs()');
+
+    my $cmd = CAF::Process->new($CMD_DPKG_QUERY, keeps_state => 1, log => $self);
+
+    my $out = $cmd->output();
+    if ($?) {
+        return 0;
+    }
+    my @pkgs = grep(m{^ii }, split(/\n/, $out));
+    for (@pkgs) {s/^...;//g}; # Strip package status characters
+
+    return Set::Scalar->new(@pkgs);
+}
+
+
+sub get_package_version_arch
+{
+    my ($self, $name, $details) = @_;
+    $self->debug(5, 'Entered get_package_version_arch()');
+
+    my @versions;
+
+    if ($details) {
+        while (my ($ver, $archs) = each(%$details)) {
+            if ($archs->{arch}) {
+                while (my $arch = each(%$archs->{arch})) {
+                    $self->debug(5, '  Adding package ', $name, ' with version ', $ver, ' and architecture ', $arch, ' to list');
+                    push(@versions, sprintf('%s:%s=%s', $name, $arch, unescape($ver)));
+                }
+            }
+            else {
+                $self->debug(5, '  Adding package ', $name, ' with version ', $ver, ' but without architecture to list');
+                push(@versions, sprintf('%s=%s', $name, unescape($ver)));
+            }
+        }
+    }
+    else {
+        $self->debug(5, '  Adding package ', $name, ' without version or architecture to list');
+        push(@versions, $name);
+    }
+
+    return \@versions;
+}
+
+
+sub apply_package_version_arch
+{
+    my ($self, $packagelist, $packagetree) = @_;
+
+    $self->debug(5, 'Entered apply_package_version_arch()');
+
+    my @packages = @$packagelist;
+    my @results;
+
+    while (my $name = shift(@packages)) {
+        my $name_escaped = escape($name);
+        if (exists($packagetree->{$name_escaped})) {
+            my $versions = $self->get_package_version_arch($name, $packagetree->{$name_escaped});
+            push(@results, @$versions);
+        }
+        else {
+            $self->error('Could not find package ', $name, ' in tree');
```

Comment:
i would collect all packages that are not found (`push*(@notfound, $name)`) , and once report one error
```perl
$self->error("Could not find packages", join(", ", @notfound)) if @notfound`
```